### ISSUE TITLE
Semantic checks for external calls

### ISSUE THREAD

**Issue by Aurel300** [flintrocks/flint#72]
Blocked by #70 and #71, some checks also blocked by #68 and #69.

Refer to the [proposal](https://github.com/flintrocks/flint/blob/master/proposals/0003-external-calls.md#implementation-requirements).

 - [ ] check that non-`@payable` functions are never invoked with a `value` hyper-parameter (#95)
 - [ ] check that `@payable` functions are invoked with a `value` hyper-parameter of type `Wei` (#95)
 - [ ] check that `if let ... = call? ...` is only used with an external function that has a declared return type (#96)
 - [ ] check that in `if let ... = call? ...` the return type matches the variable declared (#96)
 - [ ] check that `call? ...` is used only in `if let ...` constructs (#96)
 - [ ] check that `call` is in a `do ... catch` block (#97)
 - [ ] external contract member checks (#99)
 - [ ] `call` hyper-parameters checks (duplicates, only `value` and `gas` for now) (#100)
 - [ ] checks for type mapping (#90)

Any check should be supported by positive and negative tests. More checks may need to be added.

**Comment by Aurel300** [comment]
Analysis done, spawned tickets in the updated issue description.

**Comment by Aurel300** [comment]
Some tests / semantic analyser diagnostics were disabled in the #127 PR:

 - `doCatchStatementContainsNoExternalCall` - this doesn't work if the external call is not a direct descendant of the do catch block, e.g. `do { let a: Int = (call a.b()) as! Int } ...`
 - `externalCallReturnValueIgnored` - this doesn't work if the value is not assigned to a variable; it can still be used in a valid way, e.g. `if (((call a.b()) as! Int) == 0) { ... }`

The only time a value is actually not used is when it is a `.expression()` statement e.g. in a `Block`. Care should be taken with casting – even though casting technically should check if the value is in a given range, it should probably not be treated as "using" the value, especially given that at the moment we can only cast to `Int`, which does not check range (`Int === int256` is the largest integer type).

**Comment by nvgrw** [comment]
Let's create a new ticket and link to this one rather than reopening it
